shared/src/balancer_event_handler.rs

```diff
@@ -54,15 +117,127 @@ impl PoolSpecialization {
             0 => Ok(Self::General),
             1 => Ok(Self::MinimalSwapInfo),
             2 => Ok(Self::TwoToken),
-            t => Err(anyhow!("Invalid PoolSpecialization value {}", t)),
+            t => Err(anyhow!("Invalid PoolSpecialization value {} (> 2)", t)),
         }
     }
 }
 
-#[derive(Default)]
-pub struct BalancerPools {}
-
 impl BalancerPools {
+    fn try_upgrade(&mut self, pool_builder: WeightedPoolBuilder) {
+        match pool_builder.clone().try_into() {
+            Ok(weighted_pool) => {
+                let token_registration = pool_builder.tokens_registration.expect("known to exist");
+                // When upgradable, delete pending pool and add to valid pools
+                let pool_id = token_registration.pool_id;
+                tracing::info!("Upgrading Pool Builder with id {:?}", pool_id);
+                self.pending_pools.remove(&pool_id);
+                self.pools.insert(pool_id, weighted_pool);
+                for token in token_registration.tokens {
+                    self.pools_by_token
+                        .entry(token)
+                        .or_default()
+                        .insert(pool_id);
+                }
+            }
+            Err(_) => {
+                tracing::info!("Pool Builder not yet upgradable");
```

Comment:
This should probably not log at info level (which is higher signal than `debug`). I don't think we need a log entry here at all, but if it should be `debug` or `trace`.